### PR TITLE
[FIX] account: Prevent deletion of tax lines

### PR DESCRIPTION
**Pull request by smetl**
Improvement of the current fixes:
https://github.com/odoo/odoo/commit/ac3cf2ed59c1dd5a6c91da201a0d4093212b26ae
https://github.com/odoo/odoo/commit/feb4e839dc4e35ab16527e5e264b5cc1c0a76c8e

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -991,29 +998,36 @@ def _recompute_dynamic_lines(self, recompute_all_taxes=False, recompute_tax_base
         '''
         for invoice in self:
             # Dispatch lines and pre-compute some aggregated values like taxes.
+            expected_tax_rep_lines = set()
+            current_tax_rep_lines = set()
+            for line in invoice.line_ids:
+                if line.recompute_tax_line:
+                    recompute_all_taxes = True
```
Comment (william-andre) [inline addons/account/models/account_move.py:1005]: You are changing the value of `recompute_all_taxes` for all invoices in `self`

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -991,29 +998,36 @@ def _recompute_dynamic_lines(self, recompute_all_taxes=False, recompute_tax_base
...
+                    line.recompute_tax_line = False
+                if line.tax_repartition_line_id:
+                    current_tax_rep_lines.add(line.tax_repartition_line_id._origin)
+                elif line.tax_ids:
```
Comment (william-andre) [inline addons/account/models/account_move.py:1009]: Can't you have both `tax_repartition_line_id` and `tax_ids`?

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -991,29 +998,36 @@ def _recompute_dynamic_lines(self, recompute_all_taxes=False, recompute_tax_base
...
-                invoice.line_ids.recompute_tax_line = False
+            # Compute taxes.
+            if recompute_all_taxes:
                 invoice._recompute_tax_lines()
-
-            if recompute_tax_base_amount:
+            elif recompute_tax_base_amount:
                 invoice._recompute_tax_lines(recompute_tax_base_amount=True)
+            elif delta_tax_rep_lines and not self._context.get('move_reverse_cancel'):
+                invoice._recompute_tax_lines(tax_rep_lines_to_recompute=delta_tax_rep_lines)
```
Comment (william-andre) [inline addons/account/models/account_move.py:1030]: What if `recompute_tax_base_amount=True` but there are repartitions lines missing?

File: addons/account/models/account_move.py (hunk 2 above)
Comment (smetl) [reply addons/account/models/account_move.py:1009]: sure when a tax is affecting the base of the subsequent ones

File: addons/account/models/account_move.py (hunk 3 above)
Comment (smetl) [reply addons/account/models/account_move.py:1030]: Idk but the `recompute_tax_base_amount` is already hacky at the first place. I didn't want to change the current behavior

File: addons/account/models/account_move.py (hunk 2 above)
Comment (william-andre) [reply addons/account/models/account_move.py:1009]: Then `elif` doesn't work

Comment (william-andre) [reply addons/account/models/account_move.py:1010]: @smetl is the `elif` on the line below justified if you can have taxes on a repartition line?

Comment (smetl) [reply addons/account/models/account_move.py:1010]: it's ok no worry. I don't want to account the tax_ids from tax lines here

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Comment by martins-ciel** [comment]
Hi guys,

After this update I got this error when I try to create a credit note, should I open a new issue?

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/web/controllers/main.py", line 1346, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/web/controllers/main.py", line 1334, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/api.py", line 390, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/odoo/api.py", line 377, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/account/wizard/account_move_reversal.py", line 87, in reverse_moves
    new_moves = moves._reverse_moves(default_values_list, cancel=is_cancel_needed)
  File "/odoo-app/ciel-source/13.0/addons/l10n-br-ciel-it/l10n_br_ciel_it_account/models/account.py", line 400, in _reverse_moves
    reverse_moves = super(AccountMove, self)._reverse_moves(default_values_list, cancel)
  File "/odoo-app/odoo-source/13.0/enterprise/addons/account_asset/models/account_move.py", line 88, in _reverse_moves
    return super(AccountMove, self)._reverse_moves(default_values_list, cancel)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/sale/models/account_invoice.py", line 69, in _reverse_moves
    return super()._reverse_moves(default_values_list=default_values_list, cancel=cancel)
  File "/odoo-app/odoo-source/13.0/enterprise/addons/l10n_mx_edi/models/account_invoice.py", line 713, in _reverse_moves
    return super(AccountMove, self)._reverse_moves(default_values_list, cancel=cancel)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/account/models/account_move.py", line 2252, in _reverse_moves
    reverse_move._recompute_dynamic_lines(recompute_all_taxes=False)
  File "/odoo-app/odoo-source/13.0/odoo-server/addons/account/models/account_move.py", line 1031, in _recompute_dynamic_lines
    invoice._recompute_tax_lines(tax_rep_lines_to_recompute=delta_tax_rep_lines)
TypeError: _recompute_tax_lines() got an unexpected keyword argument 'tax_rep_lines_to_recompute'

**Comment by william-andre** [comment]
You are probably extending `_recompute_tax_lines` somewhere in your custom code.
You just need to add the keyword in the method signature, something like this:
```python
def _recompute_tax_lines(self, recompute_tax_base_amount=False, tax_rep_lines_to_recompute=None):
    ...
    res = super()._recompute_tax_lines(recompute_tax_base_amount, tax_rep_lines_to_recompute):
    ...
    return res
```

Sorry for the inconvenience

**Pull request closed without merging**